Add ExtendedMasterSecret support in SunPKCS11 provider

taoliult opened the pull request:
This commit is for adding the ExtendedMasterSecret support in SunPKCS11 security provider.

This is a back-port PR from JDKNext PR https://github.com/ibmruntimes/openj9-openjdk-jdk/pull/760

taoliult commented:
@keithc-ca Please help to review this back-port PR, thanks.

keithc-ca commented:
Jenkins test sanity plinux jdk11

keithc-ca commented:
Grinder for jdk_security3 failed: https://openj9-jenkins.osuosl.org/job/Grinder/3454.
```
[2024-04-09T20:54:11.601Z] java.security.ProviderException: Could not calculate PRF
[2024-04-09T20:54:11.601Z] 	at jdk.crypto.cryptoki/sun.security.pkcs11.P11TlsPrfGenerator.engineGenerateKey(P11TlsPrfGenerator.java:159)
[2024-04-09T20:54:11.601Z] 	at java.base/javax.crypto.KeyGenerator.generateKey(KeyGenerator.java:563)
[2024-04-09T20:54:11.601Z] 	at java.base/sun.security.ssl.Finished$T12VerifyDataGenerator.createVerifyData(Finished.java:314)
[2024-04-09T20:54:11.601Z] 	at java.base/sun.security.ssl.Finished$VerifyDataScheme.createVerifyData(Finished.java:201)
[2024-04-09T20:54:11.601Z] 	at java.base/sun.security.ssl.Finished$FinishedMessage.<init>(Finished.java:84)
[2024-04-09T20:54:11.601Z] 	at java.base/sun.security.ssl.Finished$T12FinishedProducer.onProduceFinished(Finished.java:392)
[2024-04-09T20:54:11.601Z] 	at java.base/sun.security.ssl.Finished$T12FinishedProducer.produce(Finished.java:379)
[2024-04-09T20:54:11.601Z] 	at java.base/sun.security.ssl.SSLHandshake.produce(SSLHandshake.java:436)
[2024-04-09T20:54:11.601Z] 	at java.base/sun.security.ssl.ServerHelloDone$ServerHelloDoneConsumer.consume(ServerHelloDone.java:182)
[2024-04-09T20:54:11.601Z] 	at java.base/sun.security.ssl.SSLHandshake.consume(SSLHandshake.java:392)
[2024-04-09T20:54:11.601Z] 	at java.base/sun.security.ssl.HandshakeContext.dispatch(HandshakeContext.java:443)
[2024-04-09T20:54:11.601Z] 	at java.base/sun.security.ssl.SSLEngineImpl$DelegatedTask$DelegatedAction.run(SSLEngineImpl.java:1076)
[2024-04-09T20:54:11.601Z] 	at java.base/sun.security.ssl.SSLEngineImpl$DelegatedTask$DelegatedAction.run(SSLEngineImpl.java:1063)
[2024-04-09T20:54:11.601Z] 	at java.base/java.security.AccessController.doPrivileged(AccessController.java:782)
[2024-04-09T20:54:11.601Z] 	at java.base/sun.security.ssl.SSLEngineImpl$DelegatedTask.run(SSLEngineImpl.java:1010)
[2024-04-09T20:54:11.601Z] 	at TestTLS12$testTLS12SunPKCS11Communication.runDelegatedTasks(TestTLS12.java:367)
[2024-04-09T20:54:11.601Z] 	at TestTLS12$testTLS12SunPKCS11Communication.run(TestTLS12.java:314)
[2024-04-09T20:54:11.601Z] 	at TestTLS12.main(TestTLS12.java:96)
[2024-04-09T20:54:11.601Z] 	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
[2024-04-09T20:54:11.601Z] 	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
[2024-04-09T20:54:11.601Z] 	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
[2024-04-09T20:54:11.601Z] 	at java.base/java.lang.reflect.Method.invoke(Method.java:572)
[2024-04-09T20:54:11.601Z] 	at com.sun.javatest.regtest.agent.MainWrapper$MainThread.run(MainWrapper.java:127)
[2024-04-09T20:54:11.601Z] 	at java.base/java.lang.Thread.run(Thread.java:839)
[2024-04-09T20:54:11.601Z] Caused by: sun.security.pkcs11.wrapper.PKCS11Exception: CKR_KEY_TYPE_INCONSISTENT
[2024-04-09T20:54:11.601Z] 	at jdk.crypto.cryptoki/sun.security.pkcs11.wrapper.PKCS11.C_SignInit(Native Method)
[2024-04-09T20:54:11.601Z] 	at jdk.crypto.cryptoki/sun.security.pkcs11.P11TlsPrfGenerator.engineGenerateKey(P11TlsPrfGenerator.java:152)
[2024-04-09T20:54:11.601Z] 	... 23 more
```

taoliult commented:
@keithc-ca 

The error is when calling native method PKCS11.C_SignInit(), the native code returns “CKR_KEY_TYPE_INCONSISTENT”.

```
[2024-04-09T20:05:32.249Z] Caused by: sun.security.pkcs11.wrapper.PKCS11Exception: CKR_KEY_TYPE_INCONSISTENT
[2024-04-09T20:05:32.249Z] 	at jdk.crypto.cryptoki/sun.security.pkcs11.wrapper.PKCS11.C_SignInit(Native Method)
```

I remember this error before in other test case when calling the same native API in NSS FIPS mode. The failure is related to the use of Hmac. It should be an existing issue which is not introduced by the EMS codes. 

And also, the EMS codes did not change any codes related to sun.security.pkcs11.wrapper.PKCS11.C_SignInit() or sun.security.pkcs11.P11TlsPrfGenerator.engineGenerateKey().

So, I think EMS updates should be good with this test failure.

keithc-ca commented:
Using the nightly build, the same tests passed: https://openj9-jenkins.osuosl.org/job/Grinder/3469.

taoliult commented:
@keithc-ca Checking this failure, will update shortly.

taoliult commented:
@keithc-ca 

The two test cases used the PRF function to derive the *master secret* from the : *pre-master secret*, "master secret" string and, client and server random values. So, from the exception, we can see it goes into the P11TlsPrfGenerator codes.

```
Caused by: sun.security.pkcs11.wrapper.PKCS11Exception: CKR_KEY_TYPE_INCONSISTENT
	at jdk.crypto.cryptoki/sun.security.pkcs11.wrapper.PKCS11.C_SignInit(Native Method)
	at jdk.crypto.cryptoki/sun.security.pkcs11.P11TlsPrfGenerator.engineGenerateKey(P11TlsPrfGenerator.java:152)
	... 23 more

```
The ExtendedMasterSecret function, enabled by default, replaces the client and server random values with a session hash. So, it won’t use the PRF way to derive the master secret. That is why these tests failed with error "CKR_KEY_TYPE_INCONSISTENT", because the key is not the correct type of key to use with the PRF mechanism. And also, I think that is why from JDK17 and above, it added the “-Djdk.tls.useExtendedMasterSecret=false” when running this test.

So, I updated the tests "sun/security/pkcs11/fips/SunJSSEKeyExchangeFIPS.java" and "sun/security/pkcs11/fips/TestTLS12.java" with “-Djdk.tls.useExtendedMasterSecret=false” when running the tests.

keithc-ca commented:
I can see that the change to `TestTLS12` amounts to back-porting the command-line from `FipsModeTLS12` in newer versions. However, I could not find anything in newer versions that corresponds to the proposed change to `SunJSSEKeyExchangeFIPS`: could you please provide a reference to support that change?

taoliult commented:
@keithc-ca 

The very beginning commit for re-introduce the test case in JDK17 is below:

https://github.com/ibmruntimes/openj9-openjdk-jdk17/commit/5ec9bf8d90a3de0c81271d6a0155e864064bed7c

It only re-introduce the TestTLS12.java test and renamed it as FipsModeTLS12.java. But it doesn't re-introduce the SunJSSEKeyExchangeFIPS.java test. And due to these two tests failed on the same codes when using the PRF function. So, I updated these two tests in JDK11 for adding the “-Djdk.tls.useExtendedMasterSecret=false”, same as what FipsModeTLS12.java did in JDK17 and above version.

keithc-ca commented:
Jenkins line endings check

keithc-ca commented:
Those tests now pass; see https://openj9-jenkins.osuosl.org/job/Grinder/3477.
